monix-eval/shared/src/main/scala/monix/eval/internal/TaskBracket.scala

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2014-2018 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.eval.internal
+
+import monix.eval.Task
+import monix.execution.misc.NonFatal
+
+private[eval] object TaskBracket {
+  /**
+    * Implementation for `Task.bracketE`.
+    */
+  def apply[A, B](
+    acquire: Task[A],
+    use: A => Task[B],
+    release: (A, Either[Option[Throwable], B]) => Task[Unit]): Task[B] = {
+
+    acquire.flatMap { a =>
+      val next = try use(a) catch { case NonFatal(e) => Task.raiseError(e) }
+      next.onCancelRaiseError(isCancel).transformWith[B](
+        b => release(a, Right(b)).map(_ => b),
```

Comment:
So as I understand that correctly, if `release` throws or return a failed task, it will shadow the exception from `use`. Maybe it should use `reportFailure` from current scheduler instead?